### PR TITLE
[FLINK-5882] [table] TableFunction (UDTF) should support variable types and variable arguments

### PR DESCRIPTION
**Pull request by clarkyzl**
Type: New Feature
Priority: Major
Components: table, udtf, TableFunction
Problem Definition: TableFunction (UDTF) should support variable types and variable arguments
Design:
1. This pull request is based on FLINK-5881
2. The only modification is on TableFunctionCallGen, and this part is similar with ScalarFunctionCallGen. It will make the the code generator accept the variable arguments without losing any part of it.
3. add some tests, both stream and dataset.

Impact Analysis: The only modification locates in TableFunctionCallGen. The orther part is based on FLINK-5881.
Test: Unit tests done.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/TableFunctionCallGen.scala [hunk 1]
```diff
@@ -45,13 +45,23 @@ class TableFunctionCallGen(
...
     // determine function signature
-    val matchingSignature = getSignature(tableFunction, signature)
+    val matchingMethod = getEvalMethod(tableFunction, signature)
       .getOrElse(throw new CodeGenException("No matching signature found."))
+    val matchingSignature = matchingMethod.getParameterTypes
+
+    // zip for variable signatures
+    var paramToOperands = matchingSignature.zip(operands)
+    var i = paramToOperands.length
+    while (i < operands.length
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/TableFunctionCallGen.scala]: I think this while loop can be replaced with `zipAll`

**Comment by clarkyzl** [comment]
Hi @KurtYoung . Thanks for your reviewing. I've pushed another commit. This patch is based on https://github.com/apache/flink/pull/3389. If we can get https://github.com/apache/flink/pull/3389 merged first, this patch will look simpler. @twalthr , could you please take a look at https://github.com/apache/flink/pull/3389 ? Thank you all so much.

**Comment by twalthr** [comment]
@clarkyzl I merged the varargs scalar functions. Can you rebase this and the Hive UDFs?

**Comment by twalthr** [comment]
Thanks @clarkyzl. Looks good. Will merge...

**Pull request closed without merging**